### PR TITLE
Fix buffer reuse

### PR DESCRIPTION
**Pull request by apelisse**
Alternative fix to issue #619. I've used the test built in #621. Not sure which one is best.

### CODE REVIEW

**Comment by apelisse** [comment]
Also, I would have to regenerate the pb.go test files

**Comment by apelisse** [comment]
OK This should be working now, I've undid the changes that I had done to XXX_Marshal, and used a different interface instead, I think this should be good.

File: proto/table_marshal.go [hunk 1]
```diff
@@ -2965,6 +2978,17 @@ func (p *Buffer) Marshal(pb Message) error {
 		if _, ok := pb.(Marshaler); ok {
 			return fmt.Errorf("proto: deterministic not supported by the Marshal method of %T", pb)
 		}
+		if _, ok := pb.(reverseMarshaler); ok {
+			return fmt.Errorf("proto: deterministic not supported by the Marshal method of %T", pb)
+		}
+
+	}
+	if m, ok := pb.(reverseMarshaler); ok {
```
Comment (jmarais) [inline proto/table_marshal.go]: Currently I think I would prefer too rather not adding another interface here and keep calling the ```XXX``` methods here.
With the direct calling we will skip the ```p.deterministic``` setting on the buffer.

Comment (jmarais) [reply proto/table_marshal.go]: I do have a safer approach here: 
https://github.com/gogo/protobuf/pull/627
Keeping the current interfaces/flow.

**Comment by jmarais** [comment]
In favor of: https://github.com/gogo/protobuf/pull/627

**Pull request closed without merging**